Libraries/LibWeb/HTML/SharedWorker.cpp

```diff
@@ -0,0 +1,184 @@
+WebIDL::ExceptionOr<GC::Ref<SharedWorker>> SharedWorker::construct_impl(JS::Realm& realm, String const& script_url, Variant<String, WorkerOptions>& options_value)
+{
+    // FIXME: 1. Let compliantScriptURL be the result of invoking the Get Trusted Type compliant string algorithm with
+    //           TrustedScriptURL, this's relevant global object, scriptURL, "SharedWorker constructor", and "script".
+    auto const& compliant_script_url = script_url;
+
+    // 2. If options is a DOMString, set options to a new WorkerOptions dictionary whose name member is set to the value
+    //    of options and whose other members are set to their default values.
+    auto options = options_value.visit(
+        [&](String& options) {
+            return WorkerOptions { .name = move(options) };
+        },
+        [&](WorkerOptions& options) {
+            return move(options);
+        });
+
+    // 3. Let outside settings be the current settings object.
+    auto& outside_settings = current_principal_settings_object();
+
+    // 4. Let urlRecord be the result of encoding-parsing a URL given compliantScriptURL, relative to outside settings.
+    auto url = outside_settings.encoding_parse_url(compliant_script_url);
+
+    // 5. If urlRecord is failure, then throw a "SyntaxError" DOMException.
+    if (!url.has_value())
+        return WebIDL::SyntaxError::create(realm, "SharedWorker constructed with invalid URL"_string);
+
+    // 7. Let outside port be a new MessagePort in outside settings's realm.
+    // NOTE: We do this first so that we can store the port as a GC::Ref.
+    auto outside_port = MessagePort::create(outside_settings.realm());
+
+    // 6. Let worker be a new SharedWorker object.
+    // 8. Assign outside port to the port attribute of worker.
+    auto worker = realm.create<SharedWorker>(realm, url.release_value(), options, outside_port);
+
+    // 9. Let callerIsSecureContext be true if outside settings is a secure context; otherwise, false.
+    auto caller_is_secure_context = HTML::is_secure_context(outside_settings);
+
+    // 10. Let outside storage key be the result of running obtain a storage key for non-storage purposes given outside settings.
+    auto outside_storage_key = StorageAPI::obtain_a_storage_key_for_non_storage_purposes(outside_settings);
+
+    // 11. Enqueue the following steps to the shared worker manager:
+    // FIXME: "A user agent has an associated shared worker manager which is the result of starting a new parallel queue."
+    //        We just use the singular global event loop for now.
+    Platform::EventLoopPlugin::the().deferred_invoke(GC::create_function(realm.heap(), [worker, outside_port, &outside_settings, caller_is_secure_context, outside_storage_key = move(outside_storage_key)]() mutable {
+        // 1. Let worker global scope be null.
+        GC::Ptr<SharedWorkerGlobalScope> worker_global_scope;
+
+        // 2. For each scope in the list of all SharedWorkerGlobalScope objects:
```

Comment:
yeah... I think this "list of all global skopes" will eventually need modeled in the UI process, but that's for later :tm: